### PR TITLE
Refine DB structure

### PR DESCRIPTION
**Pull request by 42Ar**
Refined Material fields by specializing Material class. Dropped color table, using RAL color numbers from now on. Improved helptexts and translations.

### CODE REVIEW

File: django_plastic_tickets/models.py [hunk 1]
```diff
@@ -1,13 +1,29 @@
...
+ral_color_table_source = ('django-plastic-tickets/' +
+                          'django_plastic_tickets/ral_colors.json')
+with open(ral_color_table_source) as f:
+    ral_color_table = json.load(f)
+ral_colors = {row['RAL']: row for row in ral_color_table}
+
+
+def validate_ral_number(val):
+    if val not in ral_colors:
+        raise ValidationError(gettext('This is not a valid RAL classic color number'))
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Can we have this in its own (static) class?

Comment (hd1ex) [inline django_plastic_tickets/models.py]: style: Line 26 is too long, please rewrap

File: django_plastic_tickets/models.py [hunk 2]
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
...
         return f'{self.name} ({self.production_method})'
 
 
-class MaterialColor(DescribedModel):
-    pass
+class Manufacturer(DescribedModel):
+    url = models.URLField(blank=True, null=True)
 
 
 class Material(models.Model):
```
Comment (hd1ex) [inline django_plastic_tickets/models.py:81]: Shouldn't this be abstract?

File: django_plastic_tickets/models.py [hunk 3]
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
...
+
+    def __str__(self) -> str:
+        return f'{self.name} ({self.type.name}, {self.get_color_name()})'
+
+
+class MaterialResin(Material):
+    pass
+
+
+class MaterialFilament(Material):
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: I think these names should be the other way around
-> `ResinMaterial` and `FilamentMaterial`

File: django_plastic_tickets/models.py [hunk 4]
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
...
+    nozzle_temp = models.FloatField(blank=True, null=True)
+    bed_temp = models.FloatField(blank=True, null=True)
+    diameter = models.FloatField(blank=True, null=True, default=1.75)
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: These fields need descriptions and should be next to the other ones above the methods

I expect these are still recommended optimal temps?

File: django_plastic_tickets/models.py [hunk 5]
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
...
+
+
+class MaterialPackage(models.Model):
+    """A spool, bottle or any other material package"""
+    manufacturer = models.ForeignKey(Manufacturer, on_delete=models.CASCADE)
+    empty_weight = models.FloatField(blank=True, null=True)
+    usual_capacity = models.FloatField(blank=True, null=True)
+    appearance = models.TextField()
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Please also add descriptions

File: django_plastic_tickets/models.py [hunk 6]
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
...
 
     def __str__(self) -> str:
-        return f'{self.name} ({self.type.name}, {self.color.name})'
+        return (f'{self.manufacturer}, ' +
+                f'{self.usual_capacity} g,i ' +
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: What does `g,i` mean?

File: django_plastic_tickets/models.py [hunk 7]
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
...
     label = models.PositiveIntegerField(
         help_text=gettext('The internal label to identify the material'),
         unique=True)
-    consumed = models.BooleanField(default=False)
+    initial_material_weight = models.FloatField(help_text=gettext(
+        'Initial weight of the material in gram, ' +
+        'not including the weight of the package'))
+    current_weight = models.FloatField(help_text=gettext(
+        'Current weight of the material in gram, ' +
+        'including the weight of the package'))
```
Comment (hd1ex) [inline django_plastic_tickets/models.py:180]: Why one including and one not including the weight of the package?    
Wouldn't this confusing?

File: django_plastic_tickets/models.py [hunk 8]
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
...
+    price = models.FloatField(help_text=gettext(
+        'Total price, including the package.'))
+    owner_identifier = models.TextField(
+        help_text=gettext('This should be a LDAP ID in most cases'))
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: We have the `User?-model for this purpose, please use it

File: django_plastic_tickets/models.py [hunk 9]
```diff
@@ -117,10 +181,10 @@ def get_message_row_count(self, cols=77) -> int:
 
 
 class PrintConfig(models.Model):
-    file = models.FilePathField()
+    file = models.TextField()
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Does this fix https://github.com/hd1ex/plastic-tickets/issues/1 ?

If so, please put this in another commit or, even better, in another pull request.

File: django_plastic_tickets/models.py [hunk 10]
```diff
@@ -117,10 +181,10 @@ def get_message_row_count(self, cols=77) -> int:
...
     count = models.PositiveIntegerField()
     material_type = models.ForeignKey(MaterialType, on_delete=models.CASCADE)
-    color = models.ForeignKey(MaterialColor, on_delete=models.CASCADE)
+    color = models.PositiveIntegerField()
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Please add a description for this

File: django_plastic_tickets/ral_colors.json [hunk 11]
```diff
@@ -0,0 +1 @@
...: "Graphitschwarz", "en": "Graphite black", "fr": "Noir graphite", "es": "Negro grafito", "it": "Nero grafite", "nl": "Grafietzwart"}, {"RAL": 9016, "RGB": "246-246-246", "HEX": "#F6F6F6", "de": "Verkehrswei\u00df", "en": "Traffic white", "fr": "Blanc signalisation", "es": "Blanco tr\u00e1fico", "it": "Bianco traffico", "nl": "Verkeerswit"}, {"RAL": 9017, "RGB": "030-030-030", "HEX": "#1E1E1E", "de": "Verkehrs-schwarz", "en": "Traffic black", "fr": "Noir signalisation", "es": "Negro tr\u00e1fico", "it": "Nero traffico", "nl": "Verkeerszwart"}, {"RAL": 9018, "RGB": "215-215-215", "HEX": "#D7D7D7", "de": "Papyruswei\u00df", "en": "Papyrus white", "fr": "Blanc papyrus", "es": "Blanco papiro", "it": "Bianco papiro", "nl": "Papyruswit"}, {"RAL": 9022, "RGB": "156-156-156", "HEX": "#9C9C9C", "de": "Perlhellgrau", "en": "Pearl light grey", "fr": "Gris clair nacr\u00e9", "es": "Gris claro perlado", "it": "Grigio chiaro perlato", "nl": "Parelmoer-lichtgrijs"}, {"RAL": 9023, "RGB": "130-130-130", "HEX": "#828282", "de": "Perldunkelgrau", "en": "Pearl dark grey", "fr": "Gris fon\u00e7\u00e9 nacr\u00e9", "es": "Gris oscuro perlado", "it": "Grigio scuro perlato", "nl": "Parelmoer-donkergrijs"}]
```
Comment (hd1ex) [inline django_plastic_tickets/ral_colors.json]: What is the source of this? Are we allowed to use this?    
Please add a link/comment

Can you also format this human readable with acceptable line lengths?    
Also please remember to keep the line break at the end of the file.

Lastly, this file shouldn't be located next to the python source code. Move this to another folder.

File: django_plastic_tickets/templates/plastic_tickets/ticket_view.html [hunk 12]
```diff
@@ -43,7 +43,7 @@ <h3> # Ticket {{ ticket.id }}</h3>
                 </td>
                 <td>{{ config.count }}</td>
                 <td>{{ config.material_type }}</td>
-                <td>{{ config.color }}</td>
+                <td>{{ ral_colors|get_item:config.color|get_item:"English" }}</td>
```
Comment (hd1ex) [inline django_plastic_tickets/templates/plastic_tickets/ticket_view.html]: This doesn't seem to work for me. I am getting a `None` in the ticket view.

Maybe consider adding a function/property to the model for this rather than using a filter - this should make the view simpler :wink:

File: django_plastic_tickets/templates/plastic_tickets/ticket_view.html [hunk 13]
```diff
@@ -61,4 +61,4 @@ <h3> # Ticket {{ ticket.id }}</h3>
                       wrap="soft">{{ ticket.message }}</textarea>
         </div>
     {% endif %}
-{% endblock %}
```
Comment (hd1ex) [inline django_plastic_tickets/templates/plastic_tickets/ticket_view.html:64]: Please do not remove the newline at the end of files.

File: django_plastic_tickets/views.py [hunk 14]
```diff
@@ -7,10 +7,16 @@
 from django.http import HttpRequest, HttpResponse, HttpResponseForbidden
 from django.shortcuts import render, redirect, get_object_or_404
 from django.utils.translation import gettext
+from django.template.defaulttags import register
 
 from . import models, forms, util
 
 
+@register.filter
+def get_item(dictionary, key):
```
Comment (hd1ex) [inline django_plastic_tickets/views.py]: Types are missing here

File: django_plastic_tickets/fixtures/initial_data.json [hunk 15]
```diff
@@ -33,41 +33,64 @@
...
         "fields": {
             "material": 1,
+            "package": 1,
             "label": 1,
-            "consumed": false
+            "initial_material_weight": 1000,
+            "current_weight": 523,
+            "price": 20,
+            "owner_identifier": "admin",
+            "usage_comment": "Please message the owner about the amount of Filament you have used."
```
Comment (hd1ex) [inline django_plastic_tickets/fixtures/initial_data.json]: I don't like this sample message - this text should rather be in general in some kind of a filament overview.

A better example would be `"This filament does a lot of stringing - use more retraction than usual!"`.

**Review by hd1ex — CHANGES_REQUESTED** [review]
Thanks for taking care of a model layout to enable proper filament management!

In general I like your ideas about filament management - especially using the RAL color model.    
Can you maybe also extend the description of the [color selection in the file form](https://github.com/hd1ex/django-plastic-tickets/blob/6abc87cde55ac4e21a07442667aa195df5ed2d80/django_plastic_tickets/templates/plastic_tickets/new_ticket_file_form.html#L93) with a reference to RAL?     
Maybe you know a better Website for RAL-Lookup than [Wikipedia](https://en.wikipedia.org/wiki/List_of_RAL_colors) :wink:

I added a few comments below you have to tackle before this is getting merged.    
Also please be informed that this will be the last breaking model change (updating migrations is generally a bad thing but cannot be avoided here) - so make sure this is the base we want in the end :sweat_smile: 

Lastly, here are some comments about the git workflow (maybe read the [contribution guide](https://github.com/hd1ex/plastic-tickets/blob/main/Contributing.md) again):         
- The commit message isn't up to standards and uses a different style than usual ([please read this](https://chris.beams.io/posts/git-commit/)) - also do not forget to update it, when you make further changes in this PR
- Please be prepared to rebase to apply the requested changes ([read this](https://git-rebase.io/), if you are not familiar with this concept)

Don't worry about not getting this right the first time - we are all learning these things by trying out :grin:

File: django_plastic_tickets/models.py (hunk 1 above)
Comment (42Ar) [reply django_plastic_tickets/models.py]: I will put it in a separate module.

File: django_plastic_tickets/models.py (hunk 2 above)
Comment (42Ar) [reply django_plastic_tickets/models.py:81]: To my understanding of this, if Material was abstract class then it would not appear in the database. Thus, we could not reference it from MaterialStock. We can make Material abstract, but then we would have to add a field to MaterialStock (one to reference MaterialFilament and one to reference MaterialResin) which would not be very nice. There are also other approaches, but using Multi-table inheritance here is the best option I think.

File: django_plastic_tickets/models.py (hunk 4 above)
Comment (42Ar) [reply django_plastic_tickets/models.py]: Yes these are the (manufacturer) recommended settings.

File: django_plastic_tickets/models.py (hunk 7 above)
Comment (42Ar) [reply django_plastic_tickets/models.py:180]: This might be confusing at the beginning, but you know the initial weight of the material on a filament roll or resin bottle, but you can only measure the weight of it with the package.

File: django_plastic_tickets/models.py (hunk 9 above)
Comment (42Ar) [reply django_plastic_tickets/models.py]: Yes it does. I will make a separate commit or pull request.

File: django_plastic_tickets/ral_colors.json (hunk 11 above)
Comment (42Ar) [reply django_plastic_tickets/ral_colors.json]: The source is https://gist.github.com/lunohodov/1995178 but the original source is propably https://www.ralfarbpalette.de . I assume that it is legit to use the data.

File: django_plastic_tickets/models.py (hunk 3 above)
Comment (42Ar) [reply django_plastic_tickets/models.py]: I wanted to have a common prefix for all tables which have something to do with Materials. I added human readable names for the admin fronted.

Comment (hd1ex) [reply django_plastic_tickets/models.py]: I'd rather have the names to be consistent with the admin frontend.

File: django_plastic_tickets/models.py (hunk 4 above)
Comment (hd1ex) [reply django_plastic_tickets/models.py]: Ok. Then please add this info at least to the descriptions.

File: django_plastic_tickets/models.py (hunk 2 above)
Comment (hd1ex) [reply django_plastic_tickets/models.py:81]: Ok agreeing with that

File: django_plastic_tickets/models.py (hunk 7 above)
Comment (hd1ex) [reply django_plastic_tickets/models.py:180]: I am agreeing that `current_weight` should contain the package weight but then you also have to store the **weight** of the **package** in some way, otherwise the `initial_material_weight` would solve no purpose.

I guess this was your intention with `MaterialPackage`?     
So would some inventory workflow be this:
- Add a `MaterialPackage` for a light, transparent filament spool (from Extrudr)
- Add a `Material` for brown GreenTEC from Extrudr
- For every new filament in stock create a `MaterialStock` with the weight written on the package as `initial_material_weight` and the `current_weight = empty_weight + initial_material_weight`
- Calculate the remaining weight at any time with `current_weight - empty_weight`
Can you ACK or update this example?

One side effect of this is, that we are assuming an equal weight of every package.

Another small thing I have seen is that both `Material` and `MaterialPackage` are storing a `Manufacturer`. Is this intentional?

Comment (42Ar) [reply django_plastic_tickets/models.py:180]: - Yes, this is what the typical workflow would look like.
- I think it is safe to assume that every package which looks the same has also the same weight. One can still weight a new package with material and subtract the weight of the material (which is known from the manufacturer) to check that.
- It is intentional that both `Material` and `MaterialPackage` store `Manufacturer`. This helps to distinguish the packages.

File: django_plastic_tickets/forms.py [hunk 16]
```diff
@@ -17,36 +17,29 @@
...
+    stocks = models.MaterialStock.objects.filter(label__in=stocks_lbl)
+    if len(stocks_lbl) != len(stocks):
         return False
+
     # Check for existing cached config
     config = models.PrintConfig.objects.filter(
         file=active_file, user=user).first()
 
-    # Update existing config and return
+    # Update existing config or create a new
```
Comment (hd1ex) [inline django_plastic_tickets/forms.py]: `s/or create a new/or create a new one`

File: django_plastic_tickets/forms.py [hunk 17]
```diff
@@ -17,36 +17,29 @@
...
-                                user=user,
-                                ticket=None
-                                )
+    else:
+        config = models.PrintConfig(file=active_file,
+                                    count=int(fc),
+                                    user=user,
+                                    ticket=None)
+    config.save()
+    config.material_stocks.set(stocks, clear=True)
```
Comment (hd1ex) [inline django_plastic_tickets/forms.py]: Can you reason about this change?

File: django_plastic_tickets/forms.py [hunk 18]
```diff
@@ -17,36 +17,29 @@
...
     config.save()
```
Comment (hd1ex) [inline django_plastic_tickets/forms.py]: Is this behavior (saving twice) intentional?

File: django_plastic_tickets/jinja2/plastic_tickets/ticket_mail.txt [hunk 19]
```diff
@@ -3,11 +3,14 @@
...
+    Labels of possible materials: {{ config.get_material_stock_list() }}
 {% endfor %}
 {% if message %}
 
 ----------
 
 {{ message }}
-{% endif %}
\ No newline at end of file
+{% endif %}
```
Comment (hd1ex) [inline django_plastic_tickets/jinja2/plastic_tickets/ticket_mail.txt:16]: Seems like you removed the new line here. Please put it back :D

File: django_plastic_tickets/locale/de/LC_MESSAGES/django.po [hunk 20]
```diff
@@ -18,28 +18,107 @@ msgstr ""
 "Plural-Forms: nplurals=2; plural=(n != 1);\n"
 "X-Generator: Poedit 2.4.1\n"
 
-#: django-plastic-tickets/django_plastic_tickets/forms.py:98
```
Comment (hd1ex) [inline django_plastic_tickets/locale/de/LC_MESSAGES/django.po:21]: These changes look wrong. Let's keep the project root as the origin.

File: django_plastic_tickets/migrations/0002_auto_20210126_2216.py [hunk 21]
```diff
@@ -0,0 +1,34 @@
+# Generated by Django 3.1.2 on 2021-01-26 21:16
```
Comment (hd1ex) [inline django_plastic_tickets/migrations/0002_auto_20210126_2216.py]: Please merge all migrations. There is now point in keeping them in this change, if you are redefining the model anyway.
There should only be the (new) initial one.

File: django_plastic_tickets/models.py [hunk 22]
```diff
@@ -42,63 +61,145 @@ class ProductionMethod(DescribedModel):
...
+    properties = models.ManyToManyField(MaterialProperty, blank=True)
+
+    def get_display_name(self) -> str:
+        if self.ral_color_number != 0:
+            return (f'{self.get_color_name()} '
+                    f'(RAL {self.ral_color_number})')
+        else:
+            return gettext("None")
+
+    def get_disp_list_name(self):
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Return type is missing.

File: django_plastic_tickets/models.py [hunk 23]
```diff
@@ -119,12 +220,19 @@ def get_message_row_count(self, cols=77) -> int:
...
 
+    def get_first_material(self):
+        return next(iter(self.material_stocks.all())).material
+
+    def get_color_name(self):
+        return self.get_first_material().get_display_name()
+
+    def get_material_type_name(self):
+        return self.get_first_material().get_disp_list_name()[1]
+
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Also the return types are missing here.

File: django_plastic_tickets/models.py [hunk 24]
```diff
@@ -133,3 +241,53 @@ def get_file_url(self):
...
+    def get_material_stock_list(self):
+        return ", ".join(str(s.label) for s in self.material_stocks.all())
+
+    def __str__(self):
+        if self.ticket is None:
+            tstr = "?"
+        else:
+            tstr = str(self.ticket.pk)
+        return (f'#{tstr}: {self.count}, "{self.file}",'
+                f'{self.get_material_stock_list()}')
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Types missing

File: django_plastic_tickets/models.py [hunk 25]
```diff
@@ -133,3 +241,53 @@ def get_file_url(self):
...
+
+
+def get_material_type_options(stocks: [MaterialStock]):
```
Comment (hd1ex) [inline django_plastic_tickets/models.py]: Return type missing.

File: django_plastic_tickets/templates/plastic_tickets/new_ticket_file_form.html [hunk 26]
```diff
@@ -167,10 +167,25 @@ <h4>Advanced information</h4>
...
+        update_mc_options();
+        {% if selected_color %}
+        mc.value = '{{selected_color}}';
+        {% endif %}
+        update_descriptions();
+   }
+   document.getElementById("reset-btn").onclick = reset;
+   reset();
+
     // TODO: Maybe also display description of (special) material colors
```
Comment (hd1ex) [inline django_plastic_tickets/templates/plastic_tickets/new_ticket_file_form.html]: This TODO is now done (as we now have special properties) :D

File: django_plastic_tickets/templates/plastic_tickets/ticket_view.html [hunk 27]
```diff
@@ -42,8 +42,8 @@ <h3> # Ticket {{ ticket.id }}</h3>
                     </a>
                 </td>
                 <td>{{ config.count }}</td>
-                <td>{{ config.material_type }}</td>
-                <td>{{ config.color }}</td>
+                <td>{{ config.get_material_type_name }}</td>
+                <td>{{ config.get_color_name }}</td>
```
Comment (hd1ex) [inline django_plastic_tickets/templates/plastic_tickets/ticket_view.html:49]: Maybe you also want to display the same info as in the ticket email (labels of possible materials) here.
Also consider displaying this only to the staff.

**Review by hd1ex — CHANGES_REQUESTED** [review]
This change is shaping up really well!

I really like the ideas you brought in and i am looking forward to merge this PR soon.
Here are some comments about the recent changes - most of them are cosmetic ones.

File: django_plastic_tickets/forms.py (hunk 17 above)
Comment (42Ar) [reply django_plastic_tickets/forms.py]: The PrintConfig object stores now a many to many relationship to the selected materials, because this simplifies the code above (the presence of the material must not be checked by lookup in the db, which would be unnecessary complex now due to the amount of MaterialProperties). Furthermore, the information "Materials which can be used to print the object" is now directly available and must not be calculated with extra code. Viewed externally, the user might select some material type and color but viewed internally, the user selects MaterialStocks, so it makes more sense to reference those objects than to reference the MaterialStocks properties.

File: django_plastic_tickets/forms.py (hunk 18 above)
Comment (42Ar) [reply django_plastic_tickets/forms.py]: According https://docs.djangoproject.com/en/3.1/topics/db/examples/many_to_many/ it must be saved (so that it has an id) before a many to many relationship can be set. However, maybe this only applies to new PrintConfig objects, so I have moved the first save into the else branch.

File: django_plastic_tickets/forms.py (hunk 17 above)
Comment (hd1ex) [reply django_plastic_tickets/forms.py]: :+1:

**Review by hd1ex — APPROVED** [review]
LGTM
